covid_moonshot/analysis/fah_plotting.py

```diff
@@ -1,154 +1,191 @@
-                DDG = ((results[f"solvent_fes_GEN{i}"][0]- results[f"complex_fes_GEN{i}"][0])* kT).value_in_unit(unit.kilocalories_per_mole)
-                low = ((results[f"solvent_fes_GEN{i}"][1]- results[f"complex_fes_GEN{i}"][2])* kT).value_in_unit(unit.kilocalories_per_mole)
-                high = ((results[f"solvent_fes_GEN{i}"][2]- results[f"complex_fes_GEN{i}"][1])* kT).value_in_unit(unit.kilocalories_per_mole)
-                plt.scatter(i, DDG, color='green')
-                plt.vlines(i,low,high,color='green')
-                if i > max_gen:
-                    max_gen = i
-            except KeyError:
-                continue
-
-        colors = {'solvent': 'blue', 'complex': 'red'}
-        for phase in ['solvent', 'complex']:
-            y = []
-            low = []
-            high = []
-            for i in range(n_gens):
-                try:
-                    y.append((results[f"{phase}_fes_GEN{i}"][0] * kT)
-                             .value_in_unit(unit.kilocalories_per_mole))
-                    low.append((results[f"{phase}_fes_GEN{i}"][1] * kT)
-                             .value_in_unit(unit.kilocalories_per_mole))
-                    high.append((results[f"{phase}_fes_GEN{i}"][2] * kT)
-                             .value_in_unit(unit.kilocalories_per_mole))
-                    if i > max_gen:
-                        max_gen = i
-                except KeyError:
-                    continue
-            shift = np.mean(y)
-            y = y - shift
-            low = low - shift
-            high = high - shift
-            plt.scatter([i for i in range(0, max_gen+1)],  y,
-                        color=colors[phase],label=phase)
-            for i,_ in enumerate(y):
-                plt.vlines(i,low[i],high[i],color=colors[phase])
-
-        plt.xlabel('GEN')
-        plt.ylabel('Relative free energy /'+r' kcal mol${^-1}$')
-        plt.plot([0, max_gen],
-                 [(results['binding_fe'][0]* kT).value_in_unit(unit.kilocalories_per_mole), (results['binding_fe'][0]* kT).value_in_unit(unit.kilocalories_per_mole)],
-                 color='green', linestyle=":", label='free energy (all GENS)')
-        plt.fill_between([0, max_gen],
-                         (results['binding_fe'][1]* kT).value_in_unit(unit.kilocalories_per_mole),
-                         (results['binding_fe'][2]* kT).value_in_unit(unit.kilocalories_per_mole),
-                         alpha=0.2, color='green')
-        plt.xticks([i for i in range(0, max_gen+1)])
-        plt.legend()
-        _produce_plot(f"fe_convergence_{title}")
-
-
-def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG', n_bins=100,
-                                  markers=[-2, -1, 0, 1, 2], title='Cumulative distribution',
-                                  filename='Cumulative distribution'):
+
+
+def plot_convergence(
+    gens: List[int],
+    complex_delta_fs: List[float],
+    complex_delta_f_errs: List[float],
```

Comment:
Should we move toward a data structure that encapsulates the phase (complex, solvent) and overall (binding) free energies and uncertainties in a single object?

Reply:
Yes, this already exists as a matter of fact (the `RunAnalysis` object). I intentionally kept the interface to the plotting functions low-level (i.e. not using the types in `covid_moonshot.core`) in case they would be useful to call from other code where we don't want the full baggage of the object hierarchy. I could go either way though, honestly.